src/wrapper/objects/auto_byte_array.rs

```diff
@@ -0,0 +1,95 @@
+use crate::sys::{jbyte, JNI_ABORT};
+use log::debug;
+
+use crate::{objects::JObject, JNIEnv};
+
+/// ReleaseMode
+///
+/// This defines the release mode of AutoByteArray (and AutoPrimitiveArray) resources, and
+/// related release array functions.
+#[derive(Clone, Copy)]
+#[repr(i32)]
+pub enum ReleaseMode {
+    /// Copy back the content and free the elems buffer.
+    CopyBack = 0,
+    /// Free the buffer without copying back the possible changes.
+    NoCopyBack = JNI_ABORT,
+}
+
+/// Auto-release wrapper for pointer-based byte arrays.
+///
+/// This wrapper is used to wrap pointers returned by get_byte_array_elements.
+///
+/// These arrays need to be released through a call to release_byte_array_elements.
+/// This wrapper provides automatic array release when it goes out of scope.
+pub struct AutoByteArray<'a: 'b, 'b> {
+    obj: JObject<'a>,
+    ptr: *mut jbyte,
```

Comment:
I think that `NonNull` can be used here: calling `as_mut` wouldn't require unsafe and unwrap. Additionally, I think that `AutoByteArray::new` should return `Result` because `ptr` should be null.

Reply:
I think that instead of using `NonNull`, it's simpler to change the the `release` and `commit` prototypes to directly take a raw pointer instead of a ref. Please see my next commit.

Regarding `AutoByteArray::new` returning a `Result`, do you think it's really needed? (I guess you mean 'could' instead of 'should' above). `GetByteArrayElements` will never return null (because we're passing a valid java object to it). And if somebody uses `AutoByteArray` poorly, it's their problem.

But if you really think it's better / safer, I'll change it.

Reply:
Update: No, it's not possible; `clippy` complains about raw pointer dereferences (which makes sense). OK, I'll go with the `NonNull`changes.

Reply:
Update2: Even with `ptr` being a`NonNull<jbyte>`, converting to a mutable reference still requires an `unsafe` block. So (unless I'm missing something) it´s not possible to avoid the `unsafe` blocks. In fact this is consistent with what the [NonNull](https://doc.rust-lang.org/std/ptr/struct.NonNull.html) doc says.

I can commit these changes if you want, but I don't think they add much in terms of safety. And in fact, they add a little more clutter, so I would prefer not to add them.

Reply:
You are right, I was sure that dereferencing `NonNull` was a safe operation, but it isn't. Though it wouldn't require unwrapping the return value, so personally I still think it is better. Am I missing something or are there any other difficulties except creating a `NonNull` wrapper?

As for `AutoByteArray::new` returning a `Result` - I see your point, but generally (rust) api should either be unsafe or validate its arguments. Panic instead of `Result` would also be OK, something like that (if we agree on using `NonNull`):
```rust
ptr: NonNull::new(ptr).expect("Non-null pointer expected")
```
Though I think that returning `Result`/`Option` makes api more flexible and we already have `NullPtr` error that can be used here.

Reply:
OK, I've committed the `NonNull` changes. Basically, we're trading two `unwrap`s by a `new` (and an `&mut`).

I'll now add the `Result` wrapper as a separate commit.

Reply:
Done.

These latest changes were more cosmetic than anything else in my opinion. But I agree, moving the `Result` wrapper "down the line" makes sense. And using `NonNull` adds some robustness (and helps with the consistency in the error reporting).